src/server/gc_worker/compaction_filter.rs

```diff
@@ -289,31 +285,54 @@ pub fn is_compaction_filter_allowd(cfg_value: &GcConfig, cluster_version: &Clust
 #[cfg(test)]
 pub mod tests {
     use super::*;
+    use crate::config::DbConfig;
     use crate::storage::kv::{RocksEngine as StorageRocksEngine, TestEngineBuilder};
-    use crate::storage::mvcc::tests::{must_commit, must_prewrite_delete, must_prewrite_put};
+    use crate::storage::mvcc::tests::{
+        must_commit, must_get_none, must_prewrite_delete, must_prewrite_put,
+    };
+    use engine_rocks::raw::CompactOptions;
+    use engine_rocks::util::get_cf_handle;
     use engine_rocks::RocksEngine;
-    use engine_traits::{CompactExt, Peekable, SyncMutable};
+    use engine_traits::{MiscExt, Peekable, SyncMutable};
     use txn_types::TimeStamp;
 
-    pub fn gc_by_compact(engine: &StorageRocksEngine, _: &[u8], safe_point: u64) {
-        let engine = engine.get_rocksdb();
-        // Put a new key-value pair to ensure compaction can be triggered correctly.
-        engine.put_cf("write", b"k1", b"v1").unwrap();
-        do_gc_by_compact(&engine, None, None, safe_point);
+    // Use a lock to protect concurrent compactions.
+    lazy_static! {
+        static ref LOCK: Mutex<()> = std::sync::Mutex::new(());
     }
 
     fn do_gc_by_compact(
         engine: &RocksEngine,
         start: Option<&[u8]>,
         end: Option<&[u8]>,
         safe_point: u64,
+        target_level: Option<usize>,
     ) {
+        let _guard = LOCK.lock().unwrap();
         let safe_point = Arc::new(AtomicU64::new(safe_point));
         let cfg = GcWorkerConfigManager(Arc::new(Default::default()));
         cfg.0.update(|v| v.enable_compaction_filter = true);
         let cluster_version = ClusterVersion::new(semver::Version::new(5, 0, 0));
         init_compaction_filter(engine.clone(), safe_point, cfg, cluster_version);
-        engine.compact_range("write", start, end, false, 1).unwrap();
+
+        let db = engine.as_inner();
+        let handle = get_cf_handle(db, CF_WRITE).unwrap();
+
+        let mut compact_opts = CompactOptions::new();
+        compact_opts.set_exclusive_manual_compaction(false);
+        compact_opts.set_max_subcompactions(1);
+        if let Some(target_level) = target_level {
+            compact_opts.set_change_level(true);
+            compact_opts.set_target_level(target_level as i32);
+        }
+        db.compact_range_cf_opt(handle, &compact_opts, start, end);
+    }
+
+    pub fn gc_by_compact(engine: &StorageRocksEngine, _: &[u8], safe_point: u64) {
+        let engine = RocksEngine::from_db(engine.get_rocksdb());
+        // Put a new key-value pair to ensure compaction can be triggered correctly.
+        engine.put_cf("write", b"k1", b"v1").unwrap();
```

Comment:
This value can not decode as a structure `WriteRef`. This may cause unit-test failed.

Reply:
The engine is only used in this case, and the invalid mvcc data can be handled in compaction filter.

Reply:
No.... I want to use it in another case.... Another case also need to gc all data.

Reply:
The key/value pair will be ignored by compaction filter after `keys::split_on_ts_for(key)` is called.

Reply:
Now Parse value failure won't cause it panic.